Imports

In [10]:
import pandas as pd

### Extract

Reading our Kaggle CSV file, let's look at our first 5 rows

In [11]:
stroke_df = pd.read_csv("./healthcare-dataset-stroke-data.csv")
stroke_df.head(5)

id  gender   age  hypertension  heart_disease ever_married   
0   9046    Male  67.0             0              1          Yes  \
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status   
0        Private          Urban             228.69  36.6  formerly smoked  \
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

### Data Transformation
We don't have extra or missing columns, let's take a look at our types and null counts

In [12]:
stroke_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


We have 5110 rows of data in our dataset looking through the columns we seem to only have null values in our bmi column

In [13]:
stroke_df.loc[stroke_df["bmi"].isnull()].head(5)

id  gender   age  hypertension  heart_disease ever_married   
1   51676  Female  61.0             0              0          Yes  \
8   27419  Female  59.0             0              0          Yes   
13   8213    Male  78.0             0              1          Yes   
19  25226    Male  57.0             0              1           No   
27  61843    Male  58.0             0              0          Yes   

        work_type Residence_type  avg_glucose_level  bmi smoking_status   
1   Self-employed          Rural             202.21  NaN   never smoked  \
8         Private          Rural              76.15  NaN        Unknown   
13        Private          Urban             219.84  NaN        Unknown   
19       Govt_job          Urban             217.08  NaN        Unknown   
27        Private          Rural             189.84  NaN        Unknown   

    stroke  
1        1  
8        1  
13       1  
19       1  
27       1

We would like to use bmi for later analysis, so let's replace bmi with the average depending on gender

Let's find the calculate the mean with a groupby

In [14]:
female_bmi = stroke_df.groupby("gender")["bmi"].mean().Female
male_bmi = stroke_df.groupby("gender")["bmi"].mean().Male

To find and replace null values depending on gender we'll make seperate dataframes for each gender and append together

In [15]:
female_df = stroke_df.loc[stroke_df["gender"]=="Female"].fillna(female_bmi)
male_df = stroke_df.loc[stroke_df["gender"]=="Male"].fillna(male_bmi)
stroke_df = pd.concat([female_df, male_df])
stroke_df.head(10)

id  gender   age  hypertension  heart_disease ever_married   
1   51676  Female  61.0             0              0          Yes  \
3   60182  Female  49.0             0              0          Yes   
4    1665  Female  79.0             1              0          Yes   
7   10434  Female  69.0             0              0           No   
8   27419  Female  59.0             0              0          Yes   
9   60491  Female  78.0             0              0          Yes   
10  12109  Female  81.0             1              0          Yes   
11  12095  Female  61.0             0              1          Yes   
12  12175  Female  54.0             0              0          Yes   
14   5317  Female  79.0             0              1          Yes   

        work_type Residence_type  avg_glucose_level        bmi smoking_status   
1   Self-employed          Rural             202.21  29.065758   never smoked  \
3         Private          Urban             171.23  34.400000         smokes   
4   Self-employed          Rural             174.12  24.000000   never smoked   
7         Private          Urban              94.39  22.800000   never smoked   
8         Private          Rural              76.15  29.065758        Unknown   
9         Private          Urban              58.57  24.200000        Unknown   
10        Private          Rural              80.43  29.700000   never smoked   
11       Govt_job          Rural             120.46  36.800000         smokes   
12        Private          Urban             104.51  27.300000         smokes   
14        Private          Urban             214.09  28.200000   never smoked   

    stroke  
1        1  
3        1  
4        1  
7        1  
8        1  
9        1  
10       1  
11       1  
12       1  
14       1

In [16]:
print("Columns with null values: ")
stroke_df.isnull().sum()

Columns with null values: 


id                   0
gender               0
age                  0
hypertension         0
heart_disease        0
ever_married         0
work_type            0
Residence_type       0
avg_glucose_level    0
bmi                  0
smoking_status       0
stroke               0
dtype: int64

### Looking better, no nulls

Our age column appears as a float type, but has no value in its first decimal, we can cast it back to being an int

In [17]:
stroke_df = stroke_df.astype({"age":"int"})
stroke_df.dtypes

id                     int64
gender                object
age                    int32
hypertension           int64
heart_disease          int64
ever_married          object
work_type             object
Residence_type        object
avg_glucose_level    float64
bmi                  float64
smoking_status        object
stroke                 int64
dtype: object

### Export to Analysis

In [18]:
stroke_df.to_csv("./stroke.csv")